projects/batfish/src/main/java/org/batfish/job/BatfishJobExecutor.java

```diff
@@ -41,105 +50,143 @@ public BatfishJobExecutor(
-          finishedJobs++;
-          completed.incrementAndGet();
-          finishedPercent = 100 * ((double) finishedJobs) / totalJobs;
-          JobResultT result = null;
-          try {
-            result = future.get();
-          } catch (InterruptedException | ExecutionException e) {
-            throw new BatfishException("Error executing job", e);
-          }
-          String time = CommonUtil.getTime(result.getElapsedTime());
-          Throwable failureCause = result.getFailureCause();
-          if (failureCause == null) {
-            result.applyTo(output, _logger, answerElement);
-            _logger.infof(
-                "Job terminated successfully with result: %s after elapsed time: %s - %d/%d "
-                    + "(%.1f%%) complete\n",
-                result.toString(), time, finishedJobs, totalJobs, finishedPercent);
-          } else {
-            String failureMessage =
-                "Failure running job after elapsed time: "
-                    + time
-                    + "\n-----BEGIN JOB LOG-----\n"
-                    + result
-                        .getHistory()
-                        .toString(BatfishLogger.getLogLevel(_settings.getLogLevel()))
-                    + "\n-----END JOB LOG-----";
-            BatfishException bfc = new BatfishException(failureMessage, failureCause);
-            if (_settings.getExitOnFirstError()) {
-              result.appendHistory(_logger);
-              throw bfc;
-            } else {
-              processingError = true;
-              result.appendHistory(_logger);
-              _logger.error(failureMessage + ":\n\t" + ExceptionUtils.getStackTrace(failureCause));
-              failureCauses.add(bfc);
-              if (!_haltOnProcessingError) {
-                result.applyTo(output, _logger, answerElement);
-              }
-            }
-          }
-        } else {
-          continue;
-        }
-      }
-      if (!futures.isEmpty()) {
+
+    for (int i = 0; i < jobs.size(); i++) {
+      try {
+        // getting the next completed task
+        Future<JobResultT> currentDoneFuture = completionService.take();
+        updateJobsStats();
+
+        JobResultT result = null;
         try {
-          Thread.sleep(JOB_POLLING_PERIOD_MS);
-        } catch (InterruptedException e) {
-          throw new BatfishException("interrupted while sleeping", e);
+          // getting the result of the job
+          result = currentDoneFuture.get();
+        } catch (InterruptedException | ExecutionException e) {
+          throw new BatfishException("Error executing job", e);
+        }
+        boolean jobResultError = handleJobResult(result, output, answerElement, failureCauses);
+        if (jobResultError) {
+          processingError = true;
         }
+
+      } catch (InterruptedException e) {
+        throw new BatfishException("interrupted while waiting for jobs to complete", e);
       }
     }
     pool.shutdown();
```

Comment:
need to shut down the `pool` even in case of exception.

Reply:
done, but related to above comment